lib/cache.gi

```diff
@@ -0,0 +1,46 @@
+#############################################################################
+##
+#W  cache.gi                     GAP library                 Chris Jefferson
+##
+##
+#Y  Copyright (C) 2017 University of St Andrews, Scotland
+##
+##  This file defines various types of caching data structures
+##
+##  Note that this file is read very early in GAP's startup, so we cannot
+##  make this a method and use method dispatch, or other bits of nice
+##  GAP functionality.
+##
+
+InstallGlobalFunction(MemoizePosIntFunction,
+function(func, defaults...)
+    local boundvals, original;
+
+    if LEN_LIST(defaults) = 0 then
+        original := AtomicList([]);
+    else
+        original := AtomicList(defaults[1]);
+    fi;
+
+    boundvals := DEEP_COPY_OBJ(original);
+
+    InstallMethod(FlushCaches, [],
+        function() boundvals := DEEP_COPY_OBJ(original); end);
+
+    return function(val)
+        local v;
+        # To fix: There is a race condition here in HPC-GAP
+        if IsBound(boundvals[val]) then
+            return boundvals[val];
+        else
+            # In HPCGAP, there is a race condition here,
+            # we might call 'func' twice. However, this will
+            # only cause wasted work and the alternative (locking)
+            # is expensive and would also require deciding what
+            # level of lock to use.
+            v := func(val);
+            boundvals[val] := v;
+            return v;
+        fi;
```

Comment:
Hm... shouldn't `MakeWriteOnceAtomic` be applied to `boundsval`? And then do this to avoid any races:
```gap
local bv;
bv := boundvals;
if not IsBound(bv[val]) then
  bv[val] := func(val);
fi;
return bv[val];
```
Or am I missing something?